daemon/state.go

```diff
@@ -66,86 +52,104 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
-		go func(ep *endpoint.Endpoint, wg chan<- bool) {
+	for _, ep := range possibleEPs {
+		go func(ep *endpoint.Endpoint, epRestored, epRegenerated chan<- bool) {
 			scopedLog := log.WithField(logfields.EndpointID, ep.ID)
+			if clean && !containerd.IsRunning(ep) {
+				scopedLog.Info("No workload could be associated with endpoint, removing endpoint")
+				d.deleteEndpoint(ep)
+				epRegenerated <- false
+				epRestored <- false
+				return
+			}
+
 			ep.Mutex.Lock()
+			scopedLog.Debug("Restoring endpoint")
+
 			if err := d.allocateIPsLocked(ep); err != nil {
 				ep.Mutex.Unlock()
 				scopedLog.WithError(err).Error("Failed to re-allocate IP of endpoint. Not restoring endpoint.")
 				d.deleteEndpoint(ep)
-				wg <- false
+				epRegenerated <- false
+				epRestored <- false
+				return
+			}
+
+			if d.conf.KeepConfig {
+				ep.SetDefaultOpts(nil)
+			} else {
+				ep.SetDefaultOpts(d.conf.Opts)
+				alwaysEnforce := policy.GetPolicyEnabled() == endpoint.AlwaysEnforce
+				ep.Opts.Set(endpoint.OptionPolicy, alwaysEnforce)
+			}
+
+			endpointmanager.Insert(ep)
+			epRestored <- true
+
+			if err := d.syncLabels(ep); err != nil {
+				scopedLog.WithError(err).Warn("Unable to restore endpoint")
+				ep.Mutex.Unlock()
 				return
 			}
-			ready := ep.SetStateLocked(endpoint.StateWaitingToRegenerate, "Triggering synchronous endpoint regeneration while syncing state to host")
+			ready := ep.SetStateLocked(endpoint.StateRestoring, "Triggering synchronous endpoint regeneration while syncing state to host")
 			ep.Mutex.Unlock()
+
 			if !ready {
 				scopedLog.WithField(logfields.EndpointState, ep.GetState()).Warn("Endpoint in inconsistent state")
-				wg <- false
+				epRegenerated <- false
 				return
 			}
 			if buildSuccess := <-ep.Regenerate(d, "syncing state to host"); !buildSuccess {
 				scopedLog.Warn("Failed while regenerating endpoint")
-				wg <- false
+				epRegenerated <- false
 				return
 			}
+
 			ep.Mutex.RLock()
 			scopedLog.WithField(logfields.IPAddr, []string{ep.IPv4.String(), ep.IPv6.String()}).Info("Restored endpoint with IPs")
 			ep.Mutex.RUnlock()
-			wg <- true
-		}(ep, wg)
+			epRegenerated <- true
+		}(ep, epRestored, epRegenerated)
 	}
 	endpointmanager.Mutex.Unlock()
 
-	restored, total := 0, 0
-	if nEndpoints > 0 {
-		for buildSuccess := range wg {
-			if buildSuccess {
-				restored++
-			}
-			total++
-			if total >= nEndpoints {
-				break
+	go func() {
+		restored, total := 0, 0
+		if nEndpoints > 0 {
+			for buildSuccess := range epRegenerated {
+				if buildSuccess {
+					restored++
+				}
+				total++
+				if total >= nEndpoints {
+					break
+				}
 			}
 		}
-	}
-	close(wg)
+		close(epRegenerated)
 
-	log.WithFields(log.Fields{
-		"count.restored": restored,
-		"count.total":    total,
-	}).Info("Found endpoints of which some were restored")
+		log.WithFields(log.Fields{
+			"count.restored": restored,
+			"count.total":    total,
+		}).Info("Found endpoints of which some were restored")
```

Comment:
This info should mention that it has finished regenerating all restored endpoints